File: python-perceptron/perceptron.py
class perceptron:

    def __init__(self, size: int, learn_rate: float) -> None:
        self.weights = [1. for i in range(size)]
        self.learn_rate = learn_rate

    def classify(self, values: list[float]) -> int:
        if len(values) != len(self.weights):
            raise Exception("incorrect length of values")
        sum = 0
        for i in range(len(self.weights)):
            sum += self.weights[i] * values[i]
        if sum > 0:
            return 1
        return -1

    def train(self, values: list[float], expect: int) -> None:
        if self.classify(values) != expect:
            for i in range(len(self.weights)):
                self.weights[i] = self.weights[i] + self.learn_rate * expect * values[i]

    def bulk_train(self, value_list: list[list[float]], expects: list[int]) -> None:
        if len(value_list) != len(expects):
            raise Exception("lengths of the value_list and expects do not correspond")
        for i in range(len(value_list)):
            self.train(value_list[i], expects[i])

    def evaluate(self, value_list: list[list[float]], expects: list[int]) -> int:
        if len(value_list) != len(expects):
            raise Exception("lengths of the value_list and expects do not correspond")
        wins = 0
        for i in range(len(value_list)):
            if self.classify(value_list[i]) == expects[i]:
                wins += 1
        return wins

File: java-perceptron/source/Perceptron.java
public class Perceptron {
	int size;
	float learnRate;
	float[] weights;

	public Perceptron(int size, float learnRate) {
		this.size = size;
		this.learnRate = learnRate;
		this.weights = new float[size];
		for (int i = 0; i < size; i++) {
			this.weights[i] = 1;
		}
	}

	public int classify(float[] values) throws Exception{
		if (values.length != weights.length)
			throw new Exception("incorrect length of values");
		float sum = 0;
		for (int i = 0; i < weights.length; i++)
			sum += weights[i] * values[i];
		if (sum > 0)
			return 1;
		return -1;
	}

	public void train(float[] values, int expect) throws Exception{
		if (classify(values) != expect) {
			for (int i = 0; i < weights.length; i++)
				weights[i] = weights[i] + learnRate * expect * values[i];
		}
	}

	public void bulkTrain(float[][] valueList, int[] expects) throws Exception{
		if (valueList.length != expects.length)
			throw new Exception("the lengths of valueList and expects must correspond");
		for (int i = 0; i < valueList.length; i++)
			train(valueList[i], expects[i]);
	}

	public int evaluate(float[][] valueList, int[] expects) throws Exception{
		if (valueList.length != expects.length)
			throw new Exception("the lengths of valueList and expects must correspond");
		int wins = 0;
		for (int i = 0; i < valueList.length; i++) {
			if (classify(valueList[i]) == expects[i])
				wins++;
		}
		return wins;
	}
}
